### PR TITLE
UI progress cleanup

### LINKED ISSUE #2699: add: cleanup and improve output
- [x] tidy up remaining stats (no need for `<00:00`, add total size in bytes, etc)
- [ ] add progress to other uses of `stage.dump`
- [x] profile `repo.check_modified_graph()`, `stage.save()`, `stage.dump()` etc to find out what other functions are slow, perhaps add progress for them too
  + md5sums only take about 10% of the total time. There must be other more time consuming functions
- [x] make sure we don't show an extra progress bar when there is only one target (majority of cases, see below)
- [x] cleanup output when if fails  https://github.com/iterative/dvc/issues/2699#issuecomment-548082165
- [x] large dirs progress bar hanging when collecting files before computing checksums https://github.com/iterative/dvc/issues/2770 https://github.com/iterative/dvc/issues/2699#issuecomment-548149404 
- [ ] make sure that we show HINT message to enable cache optimization properly when copy is taking too long already 

---------------------

**Extra progress bar concern:**

We started using an extra progress bar that is counting targets, as described and shown here: #2658 . In most cases though there is a single target - `dvc add file` or `dvc add directory`. That extra progress bar is not needed in those cases. It's needed in case of multiple targets or if `-R` option is specified.

### PR DESCRIPTION
**Pull request by casperdcl**
- [x] remove `<00:00`/`<??:??` from completed bars
- [x] don't use SI prefixes for small totals < 999 (i.e. `1.00/32.0 files` => `1/32 files`)
- [x] don't increment bar when a target fails (backtrack bar)
- [x] make sure we don't show an extra progress bar when there is only one target (majority of cases, see below)
- [x] large dirs progress bar hanging when collecting files before computing checksums https://github.com/iterative/dvc/issues/2770 https://github.com/iterative/dvc/issues/2699#issuecomment-548149404
- [x] cleanup output when if fails  https://github.com/iterative/dvc/issues/2699#issuecomment-548082165
- [x] add before/after recordings
- fixes #2699
- continued in #3060

**Extra progress bar concern:**

We started using an extra progress bar that is counting targets, as described and shown here: #2658 . In most cases though there is a single target - `dvc add file` or `dvc add directory`. That extra progress bar is not needed in those cases. It's needed in case of multiple targets or if `-R` option is specified.

most issues fixed:
[asciicast](https://asciinema.org/a/kuXz8YYQOoCxsQ0v3zqzOFBQE?speed=4)

large dir fix, no outer bar for single target, more progress output rather than hanging:
[asciicast](https://asciinema.org/a/1DDfK2DvD41rNk1RJGMOXM7me?speed=4)

recursive:
[asciicast](https://asciinema.org/a/hLL8k2LE7gwFoY338Io9He6ZK?speed=4)

### CODE REVIEW

**Comment by casperdcl** [comment]
@efiop could you list scenarios that you refer to for bad (too little or too much verbosity)? e.g.:

- `dvc add single_file_over_10_gb`
- `dvc add single_dir_containing_one_file_over_10_gb_and_lots_of_small_files_under_1kb`
- `dvc add single_dir_containing_many_small_files_total_over_10_gb`

Also you mentioned `RemoteBASE` so I presume `fetch/push/pull` are issues too?

**Comment by efiop** [comment]
@casperdcl Correct, I think you've listed them all. Same applies to push/pull/fetch/checkout too.

**Comment by casperdcl** [comment]
@efiop should we also consider these despite not being recommended?

- `dvc add -R single_dir_containing_many_small_files_total_over_10_gb`
- `dvc add single_dir_containing_many_small_files_total_over_10_gb/*.*`

**Comment by efiop** [comment]
@casperdcl Good point! What are your thoughts on that? 🙂

**Comment by casperdcl** [comment]
er... just wondering whether I should bother profiling them too to find out where more progress/feedback is required for users

**Comment by efiop** [comment]
@casperdcl I guess both of those cases are `dvc add small_file` times N. `dvc add small_file` is a base scenario and is handled already and `times N` part is handled by the progress bar that you've introduced earlier. So unless I'm missing something, we are already done there (or will be done indirectly).

**Comment by casperdcl** [comment]
@efiop there is much fun to be had  with that last commit (a941967371b4839f629e6a253792178624c016bb), I added progress everywhere except `check_modified_graph` - should actually catch all cases I think. Might not even want to revert any of that additional feedback.

**Comment by casperdcl** [comment]
So `stage.commit()` is a big time-consuming culprit which we hadn't added any progress for. Turns out copying files takes as long as reading & computing md5sums :upside_down_face:

**Comment by shcheklein** [comment]
A bit offtopic, sorry about that. Just to share some possible solutions to the UI when there are a few stages (top level progress, below done with [1/4], [2/4] .. ), each of stages includes a progress bar (name of the stage + remove the bar when stage is done), multiple threads progress (in this case show them all, but they are very different - probably application specific):

https://asciinema.org/a/rUwhTMR06Rz7MIFEeFxFsjrHd

**Comment by casperdcl** [comment]
@shcheklein 

> cleanup output when if fails https://github.com/iterative/dvc/issues/2699#issuecomment-548082165

when does this happen?

> make sure that we show HINT message to enable cache optimization properly when copy is taking too long already

in what cases does the hint (I assume it's a message recommending reflink caches) fail to show?

**Comment by shcheklein** [comment]
@casperdcl 

> when does this happen?

Try just some random path:

```
$ dvc add asedfsdfsd
```

> in what case(s) does the hint (I assume it's a message recommending reflink caches) fail to show?

yes, this is the "reflink" message. `cache.slow_link_warning` here https://dvc.org/doc/command-reference/config#cache 

@pared has more context about the limits we have now. You can try some huge file with a cache set to "copy". The purpose I created that checkbox for is to reconsider this policy, the way the message is written in the context of recent changes with progress bars, etc. It should be clearly visible if we hit a threshold (we can keep it the same for now?) along with progress bar (below them?).

**Comment by casperdcl** [comment]
@shcheklein right. about the reflink message, it uses `logger.warning` so plays fine with the progressbars (yes will appear above the bars). About changing the default thresholds, I think that should be a separate ticket.

**Comment by efiop** [comment]
@casperdcl Is this ready for review? Looks like it has been lost 🙁

**Comment by efiop** [comment]
@casperdcl All tests failed btw.

**Comment by casperdcl** [comment]
Not lost; will take a look this week

File: dvc/cli.py [hunk 1]
```diff
@@ -3,37 +3,39 @@
...
-import dvc.command.repro as repro
-import dvc.command.root as root
-import dvc.command.run as run
-import dvc.command.tag as tag
-import dvc.command.unprotect as unprotect
-import dvc.command.update as update
-import dvc.command.version as version
-from dvc.command.base import fix_subparsers
-from dvc.exceptions import DvcParserError
+from .command import (
```
Comment (mroutis) [inline dvc/cli.py]: are we migrating slowly to relative import?

File: dvc/progress.py [hunk 2]
```diff
@@ -58,12 +62,10 @@ def __init__(
...
         kwargs = kwargs.copy()
-        kwargs.setdefault("unit_scale", True)
         if bytes:
-            bytes_defaults = dict(
-                unit="B", unit_scale=True, unit_divisor=1024, miniters=1
-            )
-            kwargs = merge(bytes_defaults, kwargs)
+            kwargs = merge(self.BYTES_DEFAULTS, kwargs)
+        else:
+            kwargs.setdefault("unit_scale", total > 999 if total else True)
```
Comment (mroutis) [inline dvc/progress.py]: what's the reason of moving this one below an `else` clause?

File: dvc/progress.py [hunk 3]
```diff
@@ -58,12 +62,10 @@ def __init__(
         kwargs  : anything accepted by `tqdm.tqdm()`
         """
         kwargs = kwargs.copy()
-        kwargs.setdefault("unit_scale", True)
         if bytes:
-            bytes_defaults = dict(
-                unit="B", unit_scale=True, unit_divisor=1024, miniters=1
-            )
-            kwargs = merge(bytes_defaults, kwargs)
+            kwargs = merge(self.BYTES_DEFAULTS, kwargs)
```
Comment (mroutis) [inline dvc/progress.py]: If you are already changing this one:
```suggestion
            kwargs = {**self.BYTES_DEFAULTS, **kwargs}
```

File: dvc/cli.py (hunk 1 above)
Comment (casperdcl) [reply dvc/cli.py]: Yes and also fixing duplication (`import a as a` is silly)

File: dvc/progress.py (hunk 2 above)
Comment (casperdcl) [reply dvc/progress.py]: Minor efficiency. Wouldn't do anything if the `if` block above was executed.

File: dvc/repo/__init__.py [hunk 4]
```diff
@@ -25,11 +24,11 @@
...
-def locked(f):
-    @wraps(f)
+def locked(func):
+    @wraps(func)
     def wrapper(repo, *args, **kwargs):
         with repo.lock, repo.state:
-            ret = f(repo, *args, **kwargs)
+            ret = func(repo, *args, **kwargs)
             # Our graph cache is no longer valid after we release the repo.lock
             repo._reset()
```
Comment (efiop) [inline dvc/repo/__init__.py]: @casperdcl These fixes are absolutely not related to the focus this PR, right? Please try not to do those in the same PR, it gets in the way of reviewing things that matter. 🙂

File: dvc/repo/add.py [hunk 5]
```diff
@@ -65,7 +83,12 @@ def _find_all_targets(repo, target, recursive):
     if os.path.isdir(target) and recursive:
         return [
             fname
-            for fname in repo.tree.walk_files(target)
+            for fname in Tqdm(
+                repo.tree.walk_files(target),
+                desc="Recursing " + target,
```
Comment (efiop) [inline dvc/repo/add.py]: Not sure I've ever heard or seen `Recursing ` used that way :) Maybe there is a better description? `Collecting files in `? What do you think?

File: dvc/repo/add.py [hunk 6]
```diff
@@ -77,7 +100,12 @@ def _find_all_targets(repo, target, recursive):
 def _create_stages(repo, targets, fname, pbar=None):
     stages = []
 
-    for out in targets:
+    for out in Tqdm(
+        targets,
+        desc="Creating stages",
```
Comment (efiop) [inline dvc/repo/add.py]: We try not to call them "stages" in public 😄 Same as our docs, we should call them "DVC-files" instead

File: dvc/repo/add.py [hunk 7]
```diff
@@ -77,7 +100,12 @@ def _find_all_targets(repo, target, recursive):
...
+        disable=True if len(targets) < LARGE_DIR_SIZE else None,
```
Comment (efiop) [inline dvc/repo/add.py]: Is LARGE_DIR_SIZE really needed here?

File: dvc/repo/get_url.py [hunk 8]
```diff
@@ -14,6 +14,6 @@ def get_url(url, out=None):
 
     out = os.path.abspath(out)
 
-    dep, = dependency.loads_from(None, [url])
-    out, = output.loads_from(None, [out], use_cache=False)
+    (dep,) = dependency.loads_from(None, [url])
+    (out,) = output.loads_from(None, [out], use_cache=False)
```
Comment (efiop) [inline dvc/repo/get_url.py]: Again, these are unrelated changes. Let's keep them this time, but please don't mix stuff next time. Better submit a separate PR.

**Comment by efiop** [comment]
@casperdcl Would be nice to see recordings of before/after to illustrate key changes here.

File: dvc/repo/__init__.py (hunk 4 above)
Comment (casperdcl) [reply dvc/repo/__init__.py]: will ignore DeepSource/condeclimate in future

File: dvc/repo/add.py (hunk 5 above)
Comment (casperdcl) [reply dvc/repo/add.py]: Traversing?

File: dvc/repo/add.py (hunk 7 above)
Comment (casperdcl) [reply dvc/repo/add.py]: might prevent unneeded flickering (creating and disappearing) of a really quick progressbar...

File: dvc/repo/add.py (hunk 6 above)
Comment (casperdcl) [reply dvc/repo/add.py]: done

**Comment by casperdcl** [comment]
updated description with before & after recordings.
